CIPs/cip-0021.md

```diff
@@ -35,24 +40,71 @@ The core devs have been recieving complaints from the validators saying this win
 
 The proposed solution is to migrate the parameter to the EVM state (layer 2).
 
+We define the following constants:
+
+- `MIN_SAFE_LOOKBACK_WINDOW = 12`
+- `MAX_SAFE_LOOKBACK_WINDOW = 720`
+
+Before `HARDFORK_BLOCK` The `BlockchainParameters.sol` contract is updated to add a new public state variable `uptimeLookbackWindow` with a governable setter.
+
 ```solidity
   uint256 public uptimeLookbackWindow;
   event UptimeLookbackWindowSet(uint256 window);
+
+  function setUptimeLookbackWindow(uint256 window) public onlyOwner returns (uint256) {
+    require(
+      window >= MIN_SAFE_LOOKBACK_WINDOW &&window <= MAX_SAFE_LOOKBACK_WINDOW
+    );
+    ...
+  }
 ```
 
-This way, the parameter will be handled by the BlockchainParameters contract and it will be fetched with a static view call from the node (layer 1). This would mean a hard fork, replacing any reference to the parameter with a contract call.
+Starting at `HARDFORK_BLOCK`, block creation and validation MUST read the lookback window from chain state by calling the `BlockchainParameters` contract, rather than reading chain paramaters. Reward calculations are unaffected.
+
+If the return value is outside the safe window, the node MUST clamp the return value to that range such that `MIN_SAFE_LOOKBACK_WINDOW <= window <= MAX_SAFE_LOOKBACK_WINDOW`. If the result is greater than the epoch size, the node MUST discard the result and use the epoch size instead.
```

Comment:
```suggestion
If the return value is outside the safe window, the node MUST clamp the return value to that range such that `MIN_SAFE_LOOKBACK_WINDOW <= window <= MAX_SAFE_LOOKBACK_WINDOW`. If the result is greater than the epoch size, the node MUST discard the result and use `epochSize - 2` instead.
```

Reply:
Added a short comment about`epochSize` of 3 or less being invalid. If you're good with it, lets get this merged

Reply:
approved!